src/drive/web/modules/drive/Toolbar/components/Scanner.jsx

```diff
@@ -0,0 +1,210 @@
+import React from 'react'
+
+import { withClient } from 'cozy-client'
+
+import { CozyFile } from 'cozy-doctypes'
+
+class Scanner extends React.Component {
+  state = {
+    status: 'iddle',
+    error: null,
+    filename: '',
+    online: window.navigator.onLine
+  }
+  constructor(props) {
+    super(props)
+    if (!CozyFile.cozyClient) CozyFile.registerClient(this.props.client)
+  }
+
+  componentDidMount() {
+    window.addEventListener('offline', this.onOffline)
+    window.addEventListener('online', this.onOnline)
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('offline', this.onOffline)
+    window.removeEventListener('online', this.onOnline)
+  }
+
+  onOffline = () => {
+    this.setState({
+      online: false
+    })
+  }
+  onOnline = () => {
+    this.setState({
+      online: true
+    })
+  }
+  /**
+   *
+   * @param {String} imageURI native path to the file (file:///var....)
+   */
+  onSuccess = async imageURI => {
+    const {
+      generateName,
+      dirId,
+      onConflict,
+      onBeforeUpload,
+      onFinish
+    } = this.props
+    const name = generateName()
+
+    this.setState({ error: null, filename: name, status: 'uploading' })
+    /**
+     * file:/// can not be converted to a fileEntry without the Cordova's File plugin.
+     * `resolveLocalFileSystemURL` is provided by this plugin and can resolve the native
+     * path to a fileEntry readable by a `FileReader`
+     *
+     * When we finished to read the fileEntry as buffer, we start the upload process
+     *
+     */
+
+    window.resolveLocalFileSystemURL(
+      imageURI,
+      async fileEntry => {
+        fileEntry.file(
+          file => {
+            const reader = new FileReader()
+            reader.onloadend = async () => {
+              //we get the of the readAsBuffer in the `result` attr
+              try {
+                if (onBeforeUpload) onBeforeUpload()
+                const newFile = await this.uploadFileWithConflictStrategy(
+                  name,
+                  reader.result,
+                  dirId,
+                  onConflict
+                )
+                //It's possible that the filename as changed, so let's update it
+                if (onConflict === 'rename') {
+                  this.setState({ filename: newFile.data.name })
+                }
+
+                if (onFinish) onFinish()
+              } catch (error) {
+                this.setState({ error })
+              } finally {
+                this.setState({ status: 'done' })
+              }
+            }
+            // Read the file as an ArrayBuffer
+            reader.readAsArrayBuffer(file)
+          },
+          err => {
+            this.setState({ error: err })
+            this.setState({ status: 'done' })
+            console.error('error getting fileentry file!' + err)
+          }
+        )
+      },
+      error => {
+        this.setState({ error })
+        this.setState({ status: 'done' })
+      }
```

Comment:
This whole section would be clearer with separate functions instead of nested callbacks imo.